MAVProxy/modules/mavproxy_speech.py

```diff
@@ -1,26 +1,209 @@
+    speech add ALTITUDE.altitude_relative onchange every=3.0 as=altitude format=%0.0f threshold=5.0
+
+'''
+
+import time, os, math, traceback
 from MAVProxy.modules.lib import mp_module
 
+class PeriodicSpeechAnnounce(object):
+    '''
+    SpeechAnnouncement objects encapsulate all functionality and responsibility
+    of generating speech for a single mavlink message entry.
+
+    The default SpeechAnnounce module will provide mavlink-to-text and rate-limiting
+    capabilities.
+    '''
+
+    def __init__(self, module, msg_type, entry_name, sayEveryS=30.0, sayAs=None, scale=1, formatstr="%0.0f"):
+        self.module = module
+        self.msg_type = msg_type
+        self.entry_name = entry_name
+        self.sayEveryS = sayEveryS
+        self.sayAs = sayAs
+        self.scale = scale
+        self.formatstr = formatstr
+
+        self.lastTime = 0
+
+    def set_every(self, every):
+        self.sayEveryS = every
+
+    def set_as(self, say_as):
+        self.sayAs = say_as
+
+    def set_scale(self, scale):
+        self.scale = scale
+
+    def set_format(self, formatstr):
+        self.formatstr = formatstr
+
+    def match(self, msg):
+        '''Returns true if this announce is ready to handle the given message'''
+        return (msg.get_type() == self.msg_type)
+
+    def dispatch(self, msg, backend):
+        '''Dispatch this announcement IF it can handle the given message'''
+        if self.match(msg):
+            self.maybe_say(msg, backend)
+
+    def msg_entry_to_text(self, msg):
+        return getattr(msg, self.entry_name, 0)
+
+    def maybe_say(self, msg, backend):
+        '''Calls backend with this announce for the given message IF the announce passes internal checks'''
+        now = time.time()
+        if (now - self.lastTime > self.sayEveryS):
+            self.lastTime = now
+            value = self.msg_entry_to_text(msg)
+            backend(self.to_speech_text(value))
+
+    def to_speech_text(self, value):
+        '''Returns the text announced before the value'''
+        return "%s %s" % (self.get_speech_name(), self.convert(value))
+
+    def get_speech_name(self):
+        if self.sayAs:
+            return self.sayAs
+        else:
+            return self.entry_name
+
+    def convert(self, value):
+        '''Returns the announce text representation of the value'''
+        return self.formatstr % value #(value*self.scale)
+
+    def __str__(self):
+        return "%s.%s as \"%s\" every %0.0f seconds, scaling by %0.4f and formatting as %s" % (
+            self.msg_type, 
+            self.entry_name, 
+            self.get_speech_name(), 
+            self.sayEveryS,
+            self.scale,
+            self.formatstr)
+
+class ConditionalSpeechAnnounce(PeriodicSpeechAnnounce):
+    '''
+    A SpeechAnnounce subclass that provides rate-limited announcements
+    only when a mavlink value is outside an expected range.
```

Comment:
This is a subset of possible limitations you may want on announcements.

Why not allow a python expression to be supplied to determine if an announcement is warranted?